File: backend/py/src/crystalith/features/analysis/clustering.py
from __future__ import annotations

import re
from collections import Counter
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from math import sqrt

from crystalith.shared.vector_storage import VectorEntry

from .types import Topic

_STOPWORDS = {
    "a",
    "an",
    "and",
    "are",
    "as",
    "at",
    "be",
    "but",
    "by",
    "for",
    "from",
    "has",
    "have",
    "he",
    "her",
    "his",
    "if",
    "in",
    "into",
    "is",
    "it",
    "its",
    "may",
    "not",
    "of",
    "on",
    "or",
    "our",
    "she",
    "that",
    "the",
    "their",
    "them",
    "there",
    "this",
    "to",
    "was",
    "were",
    "will",
    "with",
    "you",
    "your",
}


@dataclass
class _Cluster:
    sum_vector: list[float]
    chunk_ids: list[int]

    @property
    def count(self) -> int:
        return len(self.chunk_ids)

    def centroid(self) -> list[float]:
        if not self.sum_vector or self.count == 0:
            return []
        return [value / self.count for value in self.sum_vector]

    def add(self, vector: Sequence[float], chunk_id: int) -> None:
        if not self.sum_vector:
            self.sum_vector = list(vector)
        else:
            for index, value in enumerate(vector):
                self.sum_vector[index] += value
        self.chunk_ids.append(chunk_id)


def _dot(left: Sequence[float], right: Sequence[float]) -> float:
    return sum(a * b for a, b in zip(left, right))


def _norm(vector: Sequence[float]) -> float:
    return sqrt(sum(value * value for value in vector))


def _cosine_similarity(left: Sequence[float], right: Sequence[float]) -> float:
    left_norm = _norm(left)
    right_norm = _norm(right)
    if left_norm == 0 or right_norm == 0:
        return 0.0
    return _dot(left, right) / (left_norm * right_norm)


def _extract_keywords(texts: Sequence[str], *, max_keywords: int) -> list[str]:
    tokens: list[str] = []
    for text in texts:
        for raw in re.split(r"[^A-Za-z0-9]+", text.lower()):
            cleaned = raw.strip()
            if len(cleaned) < 3:
                continue
            if cleaned in _STOPWORDS:
                continue
            tokens.append(cleaned)
    if not tokens:
        return []
    counter = Counter(tokens)
    return [token for token, _ in counter.most_common(max_keywords)]


def _topic_name(keywords: Sequence[str], fallback_index: int) -> str:
    if keywords:
        return " / ".join(keywords[:3])
    return f"Topic {fallback_index}"


def cluster_topics(
    entries: Sequence[VectorEntry],
    chunk_texts: Mapping[int, str],
    *,
    min_similarity: float = 0.7,
    max_topics: int = 10,
    max_keywords: int = 6,
) -> list[Topic]:
    if max_topics <= 0 or not entries:
        return []

    sorted_entries = sorted(entries, key=lambda entry: entry.chunk_id)
    clusters: list[_Cluster] = []

    for entry in sorted_entries:
        if not clusters:
            cluster = _Cluster(sum_vector=[], chunk_ids=[])
            cluster.add(entry.vector, entry.chunk_id)
            clusters.append(cluster)
            continue

        best_index = 0
        best_score = -1.0
        for index, cluster in enumerate(clusters):
            centroid = cluster.centroid()
            if len(centroid) != len(entry.vector):
                continue
            score = _cosine_similarity(entry.vector, centroid)
            if score > best_score:
                best_score = score
                best_index = index

        if best_score >= min_similarity or len(clusters) >= max_topics:
            clusters[best_index].add(entry.vector, entry.chunk_id)
        else:
            cluster = _Cluster(sum_vector=[], chunk_ids=[])
            cluster.add(entry.vector, entry.chunk_id)
            clusters.append(cluster)

    topics: list[Topic] = []
    for index, cluster in enumerate(clusters, start=1):
        chunk_ids = sorted(cluster.chunk_ids)
        texts = [chunk_texts.get(chunk_id, "") for chunk_id in chunk_ids]
        keywords = _extract_keywords(texts, max_keywords=max_keywords)
        topics.append(
            Topic(
                id=f"topic-{index}",
                name=_topic_name(keywords, index),
                chunk_ids=chunk_ids,
                keywords=keywords,
            )
        )

    return topics

File: apps/server/src/features/analysis/clustering.ts
// Greedy topic clustering using embedding vectors (c28 — restored from v1).
//
// Each chunk is represented by its stored embedding vector. Clusters are
// formed by greedy incremental centroid: for each chunk (sorted by id), find
// the most similar existing cluster by cosine similarity to its centroid; if
// similarity >= minSimilarity, add to that cluster, else create a new one.
//
// v1 reference: features/analysis/clustering.py (cluster_topics).
// Keyword extraction for topic naming is retained (v1 _extract_keywords).

const STOPWORDS = new Set([
  'a',
  'an',
  'and',
  'are',
  'as',
  'at',
  'be',
  'but',
  'by',
  'for',
  'from',
  'has',
  'have',
  'he',
  'her',
  'his',
  'if',
  'in',
  'into',
  'is',
  'it',
  'its',
  'may',
  'not',
  'of',
  'on',
  'or',
  'our',
  'she',
  'that',
  'the',
  'their',
  'them',
  'there',
  'this',
  'to',
  'was',
  'were',
  'will',
  'with',
  'you',
  'your',
]);

export interface Topic {
  id: string;
  name: string;
  chunkIds: number[];
  keywords: string[];
}

/** A chunk entry with its embedding vector + text (for keyword extraction). */
export interface VectorChunkEntry {
  chunkId: number;
  sourceId: number;
  vector: Float32Array;
  text: string;
}

// ---------------------------------------------------------------------------
// Vector math (v1 _dot / _norm / _cosine_similarity)
// ---------------------------------------------------------------------------

function cosineSimilarity(left: Float32Array, right: Float32Array): number {
  if (left.length !== right.length) return 0;
  let dot = 0;
  let leftNorm = 0;
  let rightNorm = 0;
  for (let i = 0; i < left.length; i++) {
    dot += left[i]! * right[i]!;
    leftNorm += left[i]! * left[i]!;
    rightNorm += right[i]! * right[i]!;
  }
  if (leftNorm === 0 || rightNorm === 0) return 0;
  return dot / (Math.sqrt(leftNorm) * Math.sqrt(rightNorm));
}

// ---------------------------------------------------------------------------
// Cluster state (v1 _Cluster: sum_vector + chunk_ids, centroid = sum / count)
// ---------------------------------------------------------------------------

interface ClusterState {
  sumVector: Float32Array;
  chunkIds: number[];
}

function newCluster(vector: Float32Array, chunkId: number): ClusterState {
  return { sumVector: new Float32Array(vector), chunkIds: [chunkId] };
}

function centroidOf(cluster: ClusterState): Float32Array {
  const count = cluster.chunkIds.length;
  const centroid = new Float32Array(cluster.sumVector.length);
  for (let i = 0; i < cluster.sumVector.length; i++) {
    centroid[i] = cluster.sumVector[i]! / count;
  }
  return centroid;
}

function addToCluster(cluster: ClusterState, vector: Float32Array, chunkId: number): void {
  for (let i = 0; i < cluster.sumVector.length; i++) {
    cluster.sumVector[i]! += vector[i]!;
  }
  cluster.chunkIds.push(chunkId);
}

// ---------------------------------------------------------------------------
// Keyword extraction for topic naming (v1 _extract_keywords + _topic_name)
// ---------------------------------------------------------------------------

function tokenize(text: string): string[] {
  const tokens: string[] = [];
  for (const raw of text.toLowerCase().split(/[^a-z0-9]+/u)) {
    if (raw.length < 3) continue;
    if (STOPWORDS.has(raw)) continue;
    tokens.push(raw);
  }
  return tokens;
}

function extractKeywords(texts: string[], maxKeywords = 6): string[] {
  const freq: Record<string, number> = {};
  for (const text of texts) {
    for (const t of tokenize(text)) {
      freq[t] = (freq[t] ?? 0) + 1;
    }
  }
  return Object.entries(freq)
    .toSorted((a, b) => b[1] - a[1])
    .slice(0, maxKeywords)
    .map(([k]) => k);
}

function topicName(keywords: string[], index: number): string {
  return keywords.length > 0 ? keywords.slice(0, 3).join(' / ') : `Topic ${index}`;
}

// ---------------------------------------------------------------------------
// clusterTopics (v1 cluster_topics — greedy incremental centroid on embeddings)
// ---------------------------------------------------------------------------

/**
 * Greedy incremental centroid clustering on embedding vectors.
 *
 * For each chunk (sorted by chunk_id), find the most similar existing cluster
 * by cosine similarity to its centroid. If >= minSimilarity (or maxTopics
 * reached), add to that cluster; otherwise create a new cluster.
 */
export function clusterTopics(
  entries: VectorChunkEntry[],
  options: { minSimilarity?: number; maxTopics?: number; maxKeywords?: number } = {},
): Topic[] {
  const { minSimilarity = 0.7, maxTopics = 10, maxKeywords = 6 } = options;
  if (maxTopics <= 0 || entries.length === 0) return [];

  const sorted = [...entries].toSorted((a, b) => a.chunkId - b.chunkId);
  const clusters: ClusterState[] = [];

  for (const entry of sorted) {
    if (clusters.length === 0) {
      clusters.push(newCluster(entry.vector, entry.chunkId));
      continue;
    }

    let bestIndex = 0;
    let bestScore = -1;

    for (let i = 0; i < clusters.length; i++) {
      const centroid = centroidOf(clusters[i]!);
      const score = cosineSimilarity(entry.vector, centroid);
      if (score > bestScore) {
        bestScore = score;
        bestIndex = i;
      }
    }

    if (bestScore >= minSimilarity || clusters.length >= maxTopics) {
      addToCluster(clusters[bestIndex]!, entry.vector, entry.chunkId);
    } else {
      clusters.push(newCluster(entry.vector, entry.chunkId));
    }
  }

  // Build topics with keyword-derived names.
  const chunkTextMap = new Map<number, string>();
  for (const e of entries) chunkTextMap.set(e.chunkId, e.text);

  return clusters.map((cluster, i) => {
    const ids = cluster.chunkIds.toSorted((a, b) => a - b);
    const texts = ids.map((id) => chunkTextMap.get(id) ?? '');
    const keywords = extractKeywords(texts, maxKeywords);
    return {
      id: `topic-${i + 1}`,
      name: topicName(keywords, i + 1),
      chunkIds: ids,
      keywords,
    };
  });
}
